File: provable-sdk-js/src/merkleProof.ts
import type {
  MerkleProofCompatibilityMismatch,
  MerkleProofCompatibilityResult,
  MerkleProofInput,
  MerkleProofLevel,
  NormalizedMerkleProof,
} from './types';
import { normalizeHexString, normalizeLevelCounts } from './proofShared';

export function normalizeMerkleProof(input: MerkleProofInput): NormalizedMerkleProof {
  if (isNormalizedMerkleProof(input)) {
    return {
      ...input,
      proof: [...input.proof],
      level_counts: [...input.level_counts],
      level_starts: [...input.level_starts],
    };
  }

  if (!input.success) {
    throw new Error(input.error || input.message || 'Missing merkle proof');
  }

  const hashItem = normalizeHexString(input.hash_item);
  const root = normalizeHexString(input.root) ?? '';
  if (!input.data_type || !hashItem || !Array.isArray(input.proof)) {
    throw new Error('Invalid merkle proof structure');
  }

  const proof = input.proof.map(hash => {
    const normalized = normalizeHexString(hash);
    if (!normalized) {
      throw new Error('Invalid proof hash');
    }
    return normalized;
  });

  return {
    data_type: input.data_type,
    hash_item: hashItem,
    proof,
    root,
    position: Number(input.position ?? 0),
    levels: Number(input.levels ?? 0),
    level_counts: (input.level_counts ?? []).map(value => Number(value)),
    level_starts: (input.level_starts ?? []).map(value => Number(value)),
    raw: input,
  };
}

export function getMerkleProofLevels(input: MerkleProofInput): MerkleProofLevel[] {
  const proof = normalizeMerkleProof(input);
  const levelCounts = normalizeLevelCounts(proof.level_counts, proof.levels, proof.proof.length);
  if (typeof levelCounts === 'string') {
    throw new Error(levelCounts);
  }

  let offset = 0;
  return levelCounts.map((count, level) => {
    const start = proof.level_starts[level] ?? defaultLevelStart(proof.position, level, count);
    const hashes = proof.proof.slice(offset, offset + count);
    offset += count;
    return {
      level,
      start,
      count,
      hashes,
    };
  });
}

export function checkMerkleProofCompatibility(
  previousInput: MerkleProofInput,
  nextInput: MerkleProofInput,
): MerkleProofCompatibilityResult {
  const previous = normalizeMerkleProof(previousInput);
  const next = normalizeMerkleProof(nextInput);
  const mismatches: MerkleProofCompatibilityMismatch[] = [];

  if (previous.data_type !== next.data_type) {
    mismatches.push({
      kind: 'data_type',
      message: `data_type mismatch previous=${previous.data_type} next=${next.data_type}`,
    });
  }

  if (previous.hash_item !== next.hash_item) {
    mismatches.push({
      kind: 'hash_item',
      message: `hash_item mismatch previous=${previous.hash_item} next=${next.hash_item}`,
      previousHash: previous.hash_item,
      nextHash: next.hash_item,
    });
  }

  if (previous.position !== next.position) {
    mismatches.push({
      kind: 'position',
      message: `position mismatch previous=${previous.position} next=${next.position}`,
      previousPosition: previous.position,
      nextPosition: next.position,
    });
  }

  const previousLevels = getMerkleProofLevels(previous);
  const nextLevels = getMerkleProofLevels(next);
  const nextLevelMaps = new Map<number, Map<number, { hash: string; index: number }>>();

  for (const level of nextLevels) {
    const positions = new Map<number, { hash: string; index: number }>();
    level.hashes.forEach((hash, index) => {
      positions.set(level.start + index, { hash, index });
    });
    nextLevelMaps.set(level.level, positions);
  }

  let checkedEntries = 0;
  for (const previousLevel of previousLevels) {
    const nextLevel = nextLevelMaps.get(previousLevel.level);
    if (!nextLevel) {
      mismatches.push({
        kind: 'missing_level',
        level: previousLevel.level,
        message: `missing level=${previousLevel.level} in new proof`,
      });
      continue;
    }

    previousLevel.hashes.forEach((previousHash, previousIndex) => {
      const position = previousLevel.start + previousIndex;
      const nextEntry = nextLevel.get(position);
      if (!nextEntry) {
        mismatches.push({
          kind: 'missing_position',
          level: previousLevel.level,
          position,
          previousIndex,
          message: `missing level=${previousLevel.level} position=${position} in new proof`,
          previousHash,
        });
        return;
      }

      checkedEntries += 1;
      if (nextEntry.hash !== previousHash) {
        mismatches.push({
          kind: 'hash_mismatch',
          level: previousLevel.level,
          position,
          previousIndex,
          nextIndex: nextEntry.index,
          previousHash,
          nextHash: nextEntry.hash,
          message: `hash mismatch level=${previousLevel.level} position=${position} previous=${previousHash} next=${nextEntry.hash}`,
        });
      }
    });
  }

  return {
    compatible: mismatches.length === 0,
    checkedEntries,
    previous,
    next,
    previousLevels,
    nextLevels,
    mismatches,
  };
}

function isNormalizedMerkleProof(input: MerkleProofInput): input is NormalizedMerkleProof {
  return 'raw' in input;
}

function defaultLevelStart(position: number, level: number, count: number): number {
  const levelPosition = positionAtLevel(position, level);
  if (count <= 0) {
    return levelPosition;
  }
  return Math.floor(levelPosition / count) * count;
}

function positionAtLevel(position: number, level: number): number {
  let current = position;
  for (let i = 0; i < level; i += 1) {
    current = Math.floor(current / 256);
  }
  return current;
}

File: provable-sdk-rs/src/merkle_proof.rs
use crate::error::{ProvableError, Result};
use crate::types::{
    MerkleProofCompatibilityMismatch, MerkleProofCompatibilityResult, MerkleProofInput,
    MerkleProofLevel, MerkleProofResponse, NormalizedMerkleProof,
};
use crate::util::normalize_hex_string;
use crate::util::normalize_level_counts;
use std::collections::HashMap;

pub fn normalize_merkle_proof(input: impl Into<MerkleProofInput>) -> Result<NormalizedMerkleProof> {
    match input.into() {
        MerkleProofInput::Normalized(value) => Ok(value),
        MerkleProofInput::Response(value) => normalize_merkle_proof_response(value),
    }
}

fn normalize_merkle_proof_response(raw: MerkleProofResponse) -> Result<NormalizedMerkleProof> {
    if !raw.success {
        return Err(ProvableError::new(
            raw.error
                .clone()
                .or(raw.message.clone())
                .unwrap_or_else(|| "Missing merkle proof".to_string()),
        ));
    }
    let hash_item = normalize_hex_string(&raw.hash_item)
        .ok_or_else(|| ProvableError::new("Invalid merkle proof structure"))?;
    let root = normalize_hex_string(&raw.root).unwrap_or_default();
    if raw.data_type.is_empty() {
        return Err(ProvableError::new("Invalid merkle proof structure"));
    }
    let proof = raw
        .proof
        .iter()
        .map(|hash| {
            normalize_hex_string(hash).ok_or_else(|| ProvableError::new("Invalid proof hash"))
        })
        .collect::<Result<Vec<_>>>()?;

    Ok(NormalizedMerkleProof {
        data_type: raw.data_type.clone(),
        hash_item,
        proof,
        root,
        position: raw.position,
        levels: raw.levels,
        level_counts: raw.level_counts.clone(),
        level_starts: raw.level_starts.clone(),
        raw,
    })
}

pub fn get_merkle_proof_levels(
    input: impl Into<MerkleProofInput>,
) -> Result<Vec<MerkleProofLevel>> {
    let proof = normalize_merkle_proof(input)?;
    let level_counts =
        normalize_level_counts(&proof.level_counts, proof.levels, proof.proof.len())?;
    let mut levels = Vec::with_capacity(level_counts.len());
    let mut offset = 0usize;
    for (level, count) in level_counts.iter().enumerate() {
        let start = proof
            .level_starts
            .get(level)
            .copied()
            .unwrap_or_else(|| default_level_start(proof.position, level, *count));
        let hashes = proof.proof[offset..offset + count].to_vec();
        levels.push(MerkleProofLevel {
            level,
            start,
            count: *count,
            hashes,
        });
        offset += count;
    }
    Ok(levels)
}

pub fn check_merkle_proof_compatibility(
    previous_input: impl Into<MerkleProofInput>,
    next_input: impl Into<MerkleProofInput>,
) -> Result<MerkleProofCompatibilityResult> {
    let previous = normalize_merkle_proof(previous_input)?;
    let next = normalize_merkle_proof(next_input)?;
    let previous_levels = get_merkle_proof_levels(previous.clone())?;
    let next_levels = get_merkle_proof_levels(next.clone())?;
    let mut mismatches = Vec::new();

    if previous.data_type != next.data_type {
        mismatches.push(MerkleProofCompatibilityMismatch {
            kind: "data_type".to_string(),
            message: format!(
                "data_type mismatch previous={} next={}",
                previous.data_type, next.data_type
            ),
            ..Default::default()
        });
    }
    if previous.hash_item != next.hash_item {
        mismatches.push(MerkleProofCompatibilityMismatch {
            kind: "hash_item".to_string(),
            previous_hash: Some(previous.hash_item.clone()),
            next_hash: Some(next.hash_item.clone()),
            message: format!(
                "hash_item mismatch previous={} next={}",
                previous.hash_item, next.hash_item
            ),
            ..Default::default()
        });
    }
    if previous.position != next.position {
        mismatches.push(MerkleProofCompatibilityMismatch {
            kind: "position".to_string(),
            previous_position: Some(previous.position),
            next_position: Some(next.position),
            message: format!(
                "position mismatch previous={} next={}",
                previous.position, next.position
            ),
            ..Default::default()
        });
    }

    let next_maps = next_levels
        .iter()
        .map(|level| {
            let map = level
                .hashes
                .iter()
                .enumerate()
                .map(|(index, hash)| (level.start + index as i64, (hash.clone(), index)))
                .collect::<HashMap<_, _>>();
            (level.level, map)
        })
        .collect::<HashMap<_, _>>();

    let mut checked_entries = 0usize;
    for level in &previous_levels {
        let Some(next_level) = next_maps.get(&level.level) else {
            mismatches.push(MerkleProofCompatibilityMismatch {
                kind: "missing_level".to_string(),
                level: Some(level.level),
                message: format!("missing level={} in new proof", level.level),
                ..Default::default()
            });
            continue;
        };

        for (previous_index, previous_hash) in level.hashes.iter().enumerate() {
            let position = level.start + previous_index as i64;
            let Some((next_hash, next_index)) = next_level.get(&position) else {
                mismatches.push(MerkleProofCompatibilityMismatch {
                    kind: "missing_position".to_string(),
                    level: Some(level.level),
                    position: Some(position),
                    previous_index: Some(previous_index),
                    previous_hash: Some(previous_hash.clone()),
                    message: format!(
                        "missing level={} position={} in new proof",
                        level.level, position
                    ),
                    ..Default::default()
                });
                continue;
            };

            checked_entries += 1;
            if next_hash != previous_hash {
                mismatches.push(MerkleProofCompatibilityMismatch {
                    kind: "hash_mismatch".to_string(),
                    level: Some(level.level),
                    position: Some(position),
                    previous_index: Some(previous_index),
                    next_index: Some(*next_index),
                    previous_hash: Some(previous_hash.clone()),
                    next_hash: Some(next_hash.clone()),
                    message: format!(
                        "hash mismatch level={} position={} previous={} next={}",
                        level.level, position, previous_hash, next_hash
                    ),
                    ..Default::default()
                });
            }
        }
    }

    Ok(MerkleProofCompatibilityResult {
        compatible: mismatches.is_empty(),
        checked_entries,
        previous,
        next,
        previous_levels,
        next_levels,
        mismatches,
    })
}

fn default_level_start(position: i64, level: usize, count: usize) -> i64 {
    let level_position = position_at_level(position, level);
    if count == 0 {
        return level_position;
    }
    (level_position / count as i64) * count as i64
}

fn position_at_level(position: i64, level: usize) -> i64 {
    let mut current = position;
    for _ in 0..level {
        current /= 256;
    }
    current
}
